agent/consul/stream/event_publisher.go

```diff
@@ -0,0 +1,305 @@
+package stream
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// EventPublisher receives changes events from Publish, and sends them to all
+// registered subscribers.
+type EventPublisher struct {
+	// topicBufferSize controls how many trailing events we keep in memory for
+	// each topic to avoid needing to snapshot again for re-connecting clients
+	// that may have missed some events. It may be zero for no buffering (the most
+	// recent event is always kept though). TODO
+	topicBufferSize int
+
+	// snapCacheTTL controls how long we keep snapshots in our cache before
+	// allowing them to be garbage collected and a new one made for subsequent
+	// requests for that topic and key. In general this should be pretty short to
+	// keep memory overhead of duplicated event data low - snapshots are typically
+	// not that expensive, but having a cache for a few seconds can help
+	// de-duplicate building the same snapshot over and over again when a
+	// thundering herd of watchers all subscribe to the same topic within a few
+	// seconds.
+	snapCacheTTL time.Duration
+
+	// This lock protects the topicBuffers, and snapCache
+	lock sync.RWMutex
+
+	// topicBuffers stores the head of the linked-list buffer to publish events to
+	// for a topic.
+	topicBuffers map[Topic]*eventBuffer
+
+	// snapCache if a cache of EventSnapshots indexed by topic and key.
+	// TODO: new struct for snapCache and snapFns and snapCacheTTL
+	snapCache map[Topic]map[string]*eventSnapshot
+
+	subscriptions *subscriptions
+
+	// publishCh is used to send messages from an active txn to a goroutine which
+	// publishes events, so that publishing can happen asynchronously from
+	// the Commit call in the FSM hot path.
+	publishCh chan changeEvents
+
+	snapshotHandlers SnapshotHandlers
+}
+
+type subscriptions struct {
+	// lock for byToken. If both subscription.lock and EventPublisher.lock need
+	// to be held, EventPublisher.lock MUST always be acquired first.
+	lock sync.RWMutex
+
+	// byToken is an mapping of active Subscriptions indexed by a the token and
+	// a pointer to the request.
+	// When the token is modified all subscriptions under that token will be
+	// reloaded.
+	// A subscription may be unsubscribed by using the pointer to the request.
+	byToken map[string]map[*SubscribeRequest]*Subscription
+}
+
+type changeEvents struct {
+	events []Event
+}
+
+// SnapshotHandlers is a mapping of Topic to a function which produces a snapshot
+// of events for the SubscribeRequest. Events are appended to the snapshot using SnapshotAppender.
+type SnapshotHandlers map[Topic]func(*SubscribeRequest, SnapshotAppender) (index uint64, err error)
+
+// SnapshotAppender appends groups of events to create a Snapshot of state.
+type SnapshotAppender interface {
+	// Append events to the snapshot.
+	// TODO: document why parameter is a slice instead of a single Event
```

Comment:
You may have figured this out but the reason is described in detail in Issue 2 of this PR description: https://github.com/hashicorp/consul/pull/7128

Reply:
Aha, perfect, thanks! That makes sense, I've encountered those problems before when polling using timestamps. I'll replace this TODO.